### PR TITLE
Enable Use of Cudnn APIs for Backward Filter Grouped Convolutions

### PR DESCRIPTION
**Pull request by AyanmoI**
In cudnn_conv_rewriter, enabling cudnn for backprop filter requires a reshape of the input shape such that it is compatible with cudnn. This reshape was missing which caused an internal error when simply trying to use cudnn backprop APIs.

The input had been reshaped in tf2xla bridge in order to use cudnn forward conv for computing gradients in case of grouped convolutions. Now that cudnn 7.6 has support for efficient backprop for grouped convolutions, there is no need to use cudnn fwd conv APIs for backprop.

Added unit-test for this change.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 1]
```diff
@@ -248,7 +246,29 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
...
+  Shape new_shape = lhs->shape();
+
+  int64 input_batch_dimension = backward_conv_dnums.input_batch_dimension();
+  int64 input_feature_dimension = backward_conv_dnums.input_feature_dimension();
+
+  int64 input_batch = new_shape.dimensions(input_batch_dimension);
+  int64 input_feature = new_shape.dimensions(input_feature_dimension);
+
+  // Ensure that input_batch is exact multiple of conv->feature_group_count()
+  CHECK_EQ(input_batch % conv->feature_group_count(), 0);
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Please output a message briefly explaining this invariant should the CHECK fail.
CHECK_EQ(...) << "..."

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 2]
```diff
@@ -248,7 +246,29 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
...
+  if (conv->feature_group_count() == 1) {
+    return std::make_tuple(true, backward_conv_window, backward_conv_dnums,
+                           lhs);
+  }
+  Shape new_shape = lhs->shape();
+
+  int64 input_batch_dimension = backward_conv_dnums.input_batch_dimension();
+  int64 input_feature_dimension = backward_conv_dnums.input_feature_dimension();
+
+  int64 input_batch = new_shape.dimensions(input_batch_dimension);
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Nit: You read and write new_shape which is a little confusing to read. What about the following:

```
int64 input_batch = lhs->shape().dimensions(input_batch_dimension);
int64 input_feature = lhs->shape().dimensions(input_feature_dimension);

Shape new_shape = lhs->shape();
new_shape.set_dimensions(input_batch_dimension,
￼                            input_batch / conv->feature_group_count());
```

**Review by thomasjoerg — DISMISSED** [review]

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Done

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc (hunk 2 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Done

**Review by thomasjoerg — CHANGES_REQUESTED** [review]
Looks like this PR breaks tests in
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/conv_depthwise_test.cc

Please have a look. Let me know if you have issues reproducing the test breakages.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 3]
```diff
@@ -258,15 +278,6 @@ MatchBackwardInput(HloInstruction* conv) {
   const auto no_match_result =
       std::make_tuple(false, Window(), ConvolutionDimensionNumbers(), nullptr);
 
-  // TODO(b/119479517): Theoretically cuDNN supports grouped convolutions also
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Can you please leave this out of this change and just work on the BackwardFilter depthwise convolution?
It would be better to have this in a separate CL so the impact can be separately checked when running our benchmarks.

**Review by akuegel — COMMENTED** [review]
In general, I think you should take a look at 
tensorflow/compiler/xla/service/convolution_group_converter.cc
If you want to map backward filter convolutions to cudnn calls, you need to remove this pass from nvptx_compiler and then handle the batch_group_count parameter, and still return no_match for feature_group_count > 1.
You can see what the bridge does for depthwise convolutions here:
tensorflow/compiler/tf2xla/kernels/conv_op_helpers.cc

**Comment by AyanmoI** [comment]
> Looks like this PR breaks tests in
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/conv_depthwise_test.cc
> 
> Please have a look. Let me know if you have issues reproducing the test breakages.

Hi Thomas.... I investigated the test failures. Now that I have enabled cudnn backward filter group convs, it looks like `MatchBackwardFilter` is incorrectly matching a forward depthwise convolution to backward filter group convolution. This check https://github.com/tensorflow/tensorflow/blob/b645595011b6a808f7186cd2fd6c4e0f21a5e281/tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc#L151 fails since the test case has a window dilation even though it is a fwd convolution, which in my opinion, is quite possible. It was passing earlier since by default any conv op with feature_group_count > 1 was returning `no_match_result`. I am not sure if I am missing something but `(input_batch_dim == output_batch_dim)` seems to be a sufficient condition to check whether a visited conv is a fwd convolution or not. Please correct me if I am wrong. I have formulated some checks to explicitly isolate depthwise fwd conv from backward filter conv, but if removing `!window_util::HasWindowDilation(conv->window())` in the matcher makes sense, the additional checks that I have formulated might be redundant. Please let me know what you think.

**Comment by AyanmoI** [comment]
> In general, I think you should take a look at
> tensorflow/compiler/xla/service/convolution_group_converter.cc
> If you want to map backward filter convolutions to cudnn calls, you need to remove this pass from nvptx_compiler and then handle the batch_group_count parameter, and still return no_match for feature_group_count > 1.
> You can see what the bridge does for depthwise convolutions here:
> tensorflow/compiler/tf2xla/kernels/conv_op_helpers.cc

Hi Adrian... I made some changes to the tf2xla bridge to  remap all depthwise backward filter convs to use feature_group_count instead of batch_group_count and use cudnn group conv BackwardFilter APIs. This was my first pass at handling depthwise backward filter and it seems to work fine. We had found that mobilent with xla enabled runs slower than tf -native. With this change xla runs faster. `convolution_group_converter` is now redundant in `nvptx_compiler` with this change. However, I understand that the other way to do this would be to remove `convolution_group_converter`in `nvptx_compiler`, add a new pass in `service/gpu` to handle batch_group convs and rewrite them to use feature_group. Which of these two would be the way to go? Making changes to the bridge might affect other backends but I don't have the complete information to make that judgment. Also, according to your last comment, I am not clear on why we would want to  "still return no_match for feature_group_count > 1" if we can transform the convs to correctly use feature_groups in a separate pass. As per my understanding we would subsequently want match these convs to backward filter conv and thunk to cudnn.  Please correct me if my understanding is wrong.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 4]
```diff
@@ -155,6 +153,15 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
                "to fold it to a backward filter convolution.";
     return no_match_result;
   }
+  auto rhs_in =
```
Comment (AyanmoI) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Although logically removing the condition `!window_util::HasWindowDilation(conv->window())` should just work, I haven't been able to completely access the scope of making that change. Hence I am a more non-invasive check that isolates depthwise fwd convs and addresses the failures. This also uses the underlying condition that for a backward filter conv, the input and the output dims cannot be the same.

**Comment by AyanmoI** [comment]
> > Looks like this PR breaks tests in
> > https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/conv_depthwise_test.cc
> > Please have a look. Let me know if you have issues reproducing the test breakages.
> 
> Hi Thomas.... I investigated the test failures. Now that I have enabled cudnn backward filter group convs, it looks like `MatchBackwardFilter` is incorrectly matching a forward depthwise convolution to backward filter group convolution. This check
> 
> https://github.com/tensorflow/tensorflow/blob/b645595011b6a808f7186cd2fd6c4e0f21a5e281/tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc#L151
> 
> fails since the test case has a window dilation even though it is a fwd convolution, which in my opinion, is quite possible. It was passing earlier since by default any conv op with feature_group_count > 1 was returning `no_match_result`. I am not sure if I am missing something but `(input_batch_dim == output_batch_dim)` seems to be a sufficient condition to check whether a visited conv is a fwd convolution or not. Please correct me if I am wrong. I have formulated some checks to explicitly isolate depthwise fwd conv from backward filter conv, but if removing `!window_util::HasWindowDilation(conv->window())` in the matcher makes sense, the additional checks that I have formulated might be redundant. Please let me know what you think.

Although removing the condition `!window_util::HasWindowDilation(conv->window())` works, I haven't been able to completely access the scope of making that change. Hence I am a more non-invasive check that isolates depthwise fwd convs and addresses the failures. This also uses the underlying condition that for a backward filter conv, the input and the output dims cannot be the same.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc (hunk 3 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: I have reverted this change and adding another separate PR

**Comment by akuegel** [comment]
> Hi Adrian... I made some changes to the tf2xla bridge to remap all depthwise backward filter convs to use feature_group_count instead of batch_group_count and use cudnn group conv BackwardFilter APIs. This was my first pass at handling depthwise backward filter and it seems to work fine. We had found that mobilenet with xla enabled runs slower than TF-native. With this change , switching backward filter to use cudnn group conv, xla performs better than TF-native. `convolution_group_converter` is now redundant in `nvptx_compiler` with this change. However, I understand that the other way to do this would be to remove `convolution_group_converter`in `nvptx_compiler`, add a new pass in `service/gpu` to handle batch_group convs and rewrite them to use feature_group. Which of these two would be the way to go? Making changes to the bridge might affect other backends but I don't have the complete information to make that judgment. Also, according to your last comment, I am not clear on why we would want to "still return no_match for feature_group_count > 1" if we can transform the convs to correctly use feature_groups in a separate pass. As per my understanding we would subsequently want match these convs to backward filter conv and thunk to cudnn. Please correct me if my understanding is wrong.

I haven't seen your changes to the bridge in this pull request, but I would expect that this would move it back towards an earlier state that was intentionally changed to introduce the batch_group_count parameter in order to be able to avoid a reshape. As far as I know, the TPU backend prefers not to have reshapes because they are not necessarily cheap.
My suggestion was meant as not changing the bridge, just look at what it does and then pattern match that. This means when you pattern match, you would never match a feature_group_count > 1 to a backward filter convolution, but instead match a batch_group_count > 1 to a backward filter convolution. Sorry for not making this clear enough :-(
Apart from that, if you actually want to pattern match this, I think you need to remove the ConvolutionFeatureGroupConverter pass from nvptx_compiler, otherwise your pattern matching will never trigger because that pass runs earlier than the CudnnConvRewriter pass. The pass is still needed for the TPU and CPU backends though.

**Comment by AyanmoI** [comment]
Hi @akuegel .. not sure why this PR got closed :(

**Comment by AyanmoI** [comment]
@akuegel anything wrong with this PR?

**Comment by AyanmoI** [comment]
> > Hi Adrian... I made some changes to the tf2xla bridge to remap all depthwise backward filter convs to use feature_group_count instead of batch_group_count and use cudnn group conv BackwardFilter APIs. This was my first pass at handling depthwise backward filter and it seems to work fine. We had found that mobilenet with xla enabled runs slower than TF-native. With this change , switching backward filter to use cudnn group conv, xla performs better than TF-native. `convolution_group_converter` is now redundant in `nvptx_compiler` with this change. However, I understand that the other way to do this would be to remove `convolution_group_converter`in `nvptx_compiler`, add a new pass in `service/gpu` to handle batch_group convs and rewrite them to use feature_group. Which of these two would be the way to go? Making changes to the bridge might affect other backends but I don't have the complete information to make that judgment. Also, according to your last comment, I am not clear on why we would want to "still return no_match for feature_group_count > 1" if we can transform the convs to correctly use feature_groups in a separate pass. As per my understanding we would subsequently want match these convs to backward filter conv and thunk to cudnn. Please correct me if my understanding is wrong.
> 
> I haven't seen your changes to the bridge in this pull request, but I would expect that this would move it back towards an earlier state that was intentionally changed to introduce the batch_group_count parameter in order to be able to avoid a reshape. As far as I know, the TPU backend prefers not to have reshapes because they are not necessarily cheap.
> My suggestion was meant as not changing the bridge, just look at what it does and then pattern match that. This means when you pattern match, you would never match a feature_group_count > 1 to a backward filter convolution, but instead match a batch_group_count > 1 to a backward filter convolution. Sorry for not making this clear enough :-(
> Apart from that, if you actually want to pattern match this, I think you need to remove the ConvolutionFeatureGroupConverter pass from nvptx_compiler, otherwise your pattern matching will never trigger because that pass runs earlier than the CudnnConvRewriter pass. The pass is still needed for the TPU and CPU backends though.

Hey Adrian...thanks for making this so clear 🙂. I didn't push my changes to this PR since I suspected that  a pass specific for  GPU backend would be ideal rather than changing the bridge. I made the bridge change to investigate how use of feature_groups rather than batch_groups would impact performance. Turns out using feature_groups and thunking to cudnn improves the performance in mobilenet (which is predominantly depthwise convs) by arnd ~20 %. I will work on this but put it in a separate PR. This PR would work without explicitly handling depthwise conv's BackwardFilter since it is being mapped to fwd convs (feature_grp = 1) anyway. I have separated the BackwardInput case [here](https://github.com/tensorflow/tensorflow/pull/30775). I have also made changes to handle test failures.

**Comment by akuegel** [comment]
Sorry, I accidentally clicked the wrong button and closed this request. I thought it meant I am finished with my review comment.
In general I am not sure if the pull request as-is adds anything. As you have seen by now, anything with feature_group_count > 1 is not a real backward filter convolution (at least it was not generated as such on the tensorflow level), and with your latest change you added one more check to make sure we don't accidentally recognize something as backward filter convolution. Does your new code still trigger, and then for which convolutions?

**Comment by AyanmoI** [comment]
The latest check I added makes sure that it accidentally doesn't match a depthwise forward convolution to a backward conv. The tests that were failing were depthwise fwd convs but with window dilation. Since an existing [check](https://github.com/tensorflow/tensorflow/blob/b645595011b6a808f7186cd2fd6c4e0f21a5e281/tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc#L151) requires a fwd convolution to not have window dilation, the testcase falsely matches the fwd conv to a backward filter conv. IMO this additional check of `!window_util::HasWindowDilation(conv->window())` [here](https://github.com/tensorflow/tensorflow/blob/b645595011b6a808f7186cd2fd6c4e0f21a5e281/tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc#L151) is most likely misleading (since fwd convs can have dilations). But I might be missing something. Since I haven't been able to investigate the complete scope of removing this additional condition (`!window_util::HasWindowDilation(conv->window())`), I added another check which isolates and identifies depthwise fwd convolutions. This additional check addresses the test failures.

**Comment by AyanmoI** [comment]
As such, this PR is to make use of cudnn backward APIs for grouped backward convolutions as they have become efficient post cudnn 7.6. There were some comments in the existing code that suggested that cudnn can be used for backward conv by simply removing the checks. However, on doing so xla threw an internal error when matching BackwardFilter. This PR enables cudnn for backward filter and fixes the  internal error. We came across this while investigating performance of ResneXt 101.

**Comment by akuegel** [comment]
> As such, this PR is to make use of cudnn backward APIs for grouped backward convolutions as they have become efficient post cudnn 7.6. There were some comments in the existing code that suggested that cudnn can be used for backward conv by simply removing the checks. However, on doing so xla threw an internal error when matching BackwardFilter. This PR enables cudnn for backward filter and fixes the internal error. We came across this while investigating performance of ResneXt 101. Also, as I mentioned earlier, I found that mobilenet's performance improves when using cudnn FilterBackprop APIs for feature_grps > 1.

The comment only refers to depthwise backward input convolution. And I have tested now that we still get regressions when we enable it. I suspect it is due to the additional reshape that we need here. I haven't looked at the generated HLO yet, but maybe there is an option for optimizing the reshape away.
Regarding backward filter convolution: as I tried to explain, the tf2xla bridge transforms depthwise backward filter convolutions to XLA convolutions with batch_group_count > 1, but feature_group_count = 1. So if you add code to try to recognize something as backward filter convolution that has feature_group_count > 1, you are likely introducing a bug here, because feature_group_count > 1 is only generated by the bridge for depthwise forward convolution and depthwise backward input convolution. You had a test failure that you then fixed by also handling one more case as forward convolution. I am interested if you would still recognize something as backward filter convolution that has feature_group_count > 1, and if yes, what convolution it is, and if you get the correct results.
There is a small chance that for some convolutions we have the choice whether to handle them as backward filter convolutions even though they are a backward input or a forward convolution at tensorflow level, but I think the right way to enable using cudnn for depthwise backward filter convolution is to implement what I suggested before, and abandon the attempt in this pull request.

**Comment by AyanmoI** [comment]
> The comment only refers to depthwise backward input convolution. And I have tested now that we still get regressions when we enable it. I suspect it is due to the additional reshape that we need here. I haven't looked at the generated HLO yet, but maybe there is an option for optimizing the reshape away.

In that case we can still continue using forward convolutions for backward input convs for feature_group_count > 1 and investigate what impact the reshape has in order to enable cudnn backward input API.

> Regarding backward filter convolution: as I tried to explain, the tf2xla bridge transforms depthwise backward filter convolutions to XLA convolutions with batch_group_count > 1, but feature_group_count = 1. So if you add code to try to recognize something as backward filter convolution that has feature_group_count > 1, you are likely introducing a bug here, because feature_group_count > 1 is only generated by the bridge for depthwise forward convolution and depthwise backward input convolution. You had a test failure that you then fixed by also handling one more case as forward convolution. I am interested if you would still recognize something as backward filter convolution that has feature_group_count > 1, and if yes, what convolution it is, and if you get the correct results.

As per my understanding, there are two cases (not completely unrelated) that tf2xla bridge handles for backward filter when feature_grps >:
1)  When feature_groups > 1 but the conv is NOT depthwise - In this case, the xla conv op it creates has feature_group_count > 1 but batch_group_count = 1. The backward filter operation  is done using a fwd conv as described [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_grad_ops.h). The input also gets transformed [here](https://github.com/tensorflow/tensorflow/blob/e1f6d9e43de2227e3adf038f538cd88cae37ff06/tensorflow/compiler/tf2xla/kernels/conv_op_helpers.cc#L516).
2) When feature_group_count > 1 and the conv IS depthwise - In this case, the xla conv op has feature_group_count = 1 but batch_group_count > 1. Batch_group convolutions are leveraged to implement backward filter for depthwise convs. Note that, unlike in case (1),  the input is not transformed here. Only the dims are swaped as required by the algorithm (same as case (1))

My change here only handles case (1). The backward filter convs that get matched have feature_group_count > 1 but are not depthwise. This is so because, as it stand now, no depthwise filter backprop will have feature_group_count > 1 and hence my change won't even be triggered. They will still be computed using forward convs. So in short, to answer your question, yes, backward filter convolution that have feature_group_count > 1 will be correctly matched but only if they are NOT depthwise. 

> There is a small chance that for some convolutions we have the choice whether to handle them as backward filter convolutions even though they are a backward input or a forward convolution at tensorflow level, but I think the right way to enable using cudnn for depthwise backward filter convolution is to implement what I suggested before, and abandon the attempt in this pull request.

You are right...we should handle depthwise backward filter convs through a different pass but we still need this PR. This so because this PR will handle(as in enable use of cudnn) for all backward filters with feature_group_count >1 but are not depthwise. I plan to add a pass to identify specifically depthwise backward filter conv (only these will have batch_group > 1 and feature_group = 1 based on what the bridge does) and replace them with convs with equivalent convs with feature_group_count > 1 and batch_group =1 so that now cudnn_conv_rewriter.cc matches them what I described in case (1) i.e, regular backward filter with feature_group > 1. So my plan is to have another PR to handle depthwise backward filter (i.e, write a pass to map them to regular backward grouped filter convs) but this PR in itself is fine and necessary to handle all other cases. I hope I haven't rambled too much :P . Please let me know if this makes sense or if I am missing something.

**Comment by akuegel** [comment]
Thanks for the explanation and linking to the bridge code line. You are right, there are cases with feature_group_count > 1 which are backward filter convolutions. Probably this happens for the more general case of grouped convolutions (not depthwise). Now I see why this PR is needed in addition to what I suggested. Thanks for working on this :)

**Review by akuegel — DISMISSED** [review]

**Comment by AyanmoI** [comment]
@thomasjoerg I have addressed the test failures...Can you please take a look :)

**Comment by akuegel** [comment]
> @thomasjoerg I have addressed the test failures...Can you please take a look :)

It turns out there are still two test failures in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/conv_ops_test.py
Conv2DTest.testInputGradientGroupConv 
Conv2DTest.testFilterGradientGroupConv
The latter should definitely work, this is not a case of matching the wrong thing. I already asked before if you have a specific convolution where you have verified that the result is correct which makes use of the cudnn call to grouped backward filter convolution. I haven't thought myself about what is needed, but it looks like just doing a reshape is not enough.
Can you please try to debug this? Conv2DTest.testFilterGradientGroupConv should definitely be passing by using your new code.

**Comment by AyanmoI** [comment]
> > @thomasjoerg I have addressed the test failures...Can you please take a look :)
> 
> It turns out there are still two test failures in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/conv_ops_test.py
> Conv2DTest.testInputGradientGroupConv
> Conv2DTest.testFilterGradientGroupConv
> The latter should definitely work, this is not a case of matching the wrong thing. I already asked before if you have a specific convolution where you have verified that the result is correct which makes use of the cudnn call to grouped backward filter convolution. I haven't thought myself about what is needed, but it looks like just doing a reshape is not enough.
> Can you please try to debug this? Conv2DTest.testFilterGradientGroupConv should definitely be passing by using your new code.
> Edit: I forgot to say, when running the test, make sure you have cuda enabled. I am not sure how that works in the open source build, but internally we use --config=cuda for that. Also, you need the _xla_gpu suffix after the test target name.

Thank Adrian for pointing this out :). Indeed I was missing a transpose due to the way that bridge handles this. I tried to follow the existing `MatchBackwardInput` but turns out the existing code in `MatchBackwardInput` also has a bug has a similar bug. I fixed the issue and will push the changes soon. In the meantime, are there any other tests that I shud be running to qualify?  Are all other tests  passing? The CI looks clean. I am not sure which checks show test failures :\

**Comment by AyanmoI** [comment]
@akuegel I pushed the changes...please take a look.. I ran all the tests in kernel_tests and they are clean. There were some failures (some sporadic) but were unrelated to my changes and independent of whether xla is enabled. All conv tests are clean.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 5]
```diff
@@ -266,7 +329,7 @@ MatchBackwardInput(HloInstruction* conv) {
   if (conv->feature_group_count() > 1) {
     return no_match_result;
   }
-
+
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: nit: it seems this introduces some extra whitespace

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 6]
```diff
@@ -155,6 +153,15 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
                "to fold it to a backward filter convolution.";
     return no_match_result;
   }
+  auto rhs_in =
+      conv->mutable_operand(1)->shape().dimensions(kernel_input_feature_dim);
+  if ((conv->feature_group_count() > 1) && (rhs_in == 1) &&
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: nit: please remove redundant brackets

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 7]
```diff
@@ -248,7 +255,63 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
...
+  int64 input_batch_dimension = backward_conv_dnums.input_batch_dimension();
+  int64 input_feature_dimension = backward_conv_dnums.input_feature_dimension();
+
+  int64 input_batch = lhs->shape().dimensions(input_batch_dimension);
+  int64 input_feature = lhs->shape().dimensions(input_feature_dimension);
+
+  // Reshape batch_dim G*N -> [G,N]
+  std::vector<int64> reshape_dims = lhs->shape().dimensions();
+  auto num_groups = conv->feature_group_count();
+  // Ensure that input_batch is exact multiple of conv->feature_group_count()
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Although in general comments are good, in this case I think the line of code below says the same thing.
So please either make the comment explain why we ensure it, or remove the comment.

File: tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc [hunk 8]
```diff
@@ -248,7 +255,63 @@ std::tuple<bool, Window, ConvolutionDimensionNumbers> MatchBackwardFilter(
...
+  CHECK_EQ(input_batch % conv->feature_group_count(), 0)
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc]: Please use the num_groups variable instead of writing conv->feature_group_count() again.

**Comment by akuegel** [comment]
> Thanks Adrian for pointing this out :). Indeed I was missing a transpose due to the way that bridge handles this. I tried to follow the existing `MatchBackwardInput` in `cudnn_conv_rewriter` but turns out the existing code in `MatchBackwardInput` also has a similar bug. I fixed the issue and will push the changes soon. In the meantime, are there any other tests that I shud be running to qualify? Are all other tests passing? The CI looks clean. I am not sure which checks show test failures :\
Yes, I think those were the only failing tests. If those pass now, it is a very good sign and things should be ok. MatchBackwardInput might have the same bug, but I thought it should work for depthwise convolution. It doesn't handle the grouped convolution case which probably requires something more along the lines what you did here. But I also locally debugged that for the grouped convolution backward input, we would never patter match it right now (at least not for the test cases that we have).

**Review by akuegel — APPROVED** [review]
Thanks a lot for working on this :)
Lets give this another try, I am hopeful that this time it will work.

**Comment by AyanmoI** [comment]
> I am hopeful that this time it will work.

:) Hopefully covered all the bases

**Comment by AyanmoI** [comment]
@akuegel I see an `import/copybara` failure...not sure what it is about...Do I need any changes on my end?

**Comment by akuegel** [comment]
> @akuegel I see an `import/copybara` failure...not sure what it is about...Do I need any changes on my end?

I actually don't know. Maybe it is because there exists still something from the previous import attempt? I deleted that changelist now, but I don't know how to retrigger  another import.

**Comment by akuegel** [comment]
Ok, a team mate showed me how to import it manually. I am doing this now.
